lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,336 @@
+/// in the same as calling `DateTime(996)`.
+class FixedDateTimeFormatter {
+  static const _powersOfTen = [1, 10, 100, 1000, 10000, 100000];
+  static const _validFormatCharacters = [
+    _yearCode,
+    _monthCode,
+    _dayCode,
+    _decadeCode,
+    _centuryCode,
+    _hourCode,
+    _minuteCode,
+    _secondCode,
+    _fractionSecondCode,
+  ];
+  static const _yearCode = 0x59; /*Y*/
+  static const _monthCode = 0x4D; /*M*/
+  static const _dayCode = 0x44; /*D*/
+  static const _decadeCode = 0x45; /*E*/
+  static const _centuryCode = 0x43; /*C*/
+  static const _hourCode = 0x68; /*H*/
+  static const _minuteCode = 0x6D; /*m*/
+  static const _secondCode = 0x73; /*s*/
+  static const _fractionSecondCode = 0x53; /*S*/
+
+  /// The format pattern string of this formatter, stored publicly in case the
+  /// user wants to retrieve it.
+  final String pattern;
+
+  /// Whether to create UTC [DateTime] objects when parsing.
+  ///
+  /// If not, the created [DateTime] objects are in the local time zone.
+  final bool isUtc;
+
+  final _blocks = _ParsedFormatBlocks();
+
+  /// Creates a new [FixedDateTimeFormatter] with the provided [pattern].
+  ///
+  /// The [pattern] interprets the characters mentioned in
+  /// [FixedDateTimeFormatter] to represent fields of a `DateTime` value. Other
+  /// characters are not special. If [isUtc] is set to false, the DateTime is
+  /// constructed with respect to the local timezone.
+  ///
+  /// There must at most be one sequence of each special character to ensure a
+  /// single source of truth when constructing the [DateTime], so a pattern of
+  /// `"CCCC-MM-DD, CC"` is invalid, because it has two separate `C` sequences.
+  FixedDateTimeFormatter(this.pattern, {this.isUtc = true}) {
+    int? currentCharacter;
+    var start = 0;
+    var characters = pattern.codeUnits;
+    for (var i = 0; i < characters.length; i++) {
+      var formatCharacter = characters[i];
+      if (currentCharacter != formatCharacter) {
+        _blocks.saveBlock(currentCharacter, start, i);
+        if (_validFormatCharacters.contains(formatCharacter)) {
+          var hasSeenBefore = _blocks.formatCharacters.indexOf(formatCharacter);
+          if (hasSeenBefore > -1) {
+            throw FormatException(
+                "Pattern contains more than one '$formatCharacter' block.\n"
+                "Previous occurrence at index ${_blocks.starts[hasSeenBefore]}",
+                pattern,
+                i);
+          } else {
+            start = i;
+            currentCharacter = formatCharacter;
+          }
+        } else {
+          currentCharacter = null;
+        }
+      }
+    }
+    _blocks.saveBlock(currentCharacter, start, pattern.length);
+  }
+
+  /// Converts a [DateTime] to a [String] as specified by the [pattern].
+  ///
+  /// Throws a [FormatException] if trying to encode a negative year.
+  String encode(DateTime datetime) {
```

Comment:
We use `dateTime` in other places. Be consistent in capitalization, probably with the upper-case `T` to match the class.